language/resource_types.md

```diff
@@ -0,0 +1,502 @@
+# Puppet Resource Types
+
+This document describes puppet's Ruby DSL for defining resource types, aka [custom types](https://docs.puppet.com/guides/custom_types.html#types). Examples are taken from the `puppetlabs-aws` module and core puppet.
+
+The second part of the document describes issues that cause puppet resource types to leak across environments (lack of environment isolation).
+
+## Puppet::Type.newtype method
+
+Defines a new puppet resource type and registers the type using the specified symbolic name. For example, to define a resource type named `ec2_instance`: 
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  # definition goes here
+end
+```
+
+The `newtype` method creates a new class `Puppet::Type::Ec2_instance`, which subclasses `Puppet::Type`. The block passed to the `newtype` method defines methods and variables on the `Puppet::Type::Ec2_instance` class.
```

Comment:
This seems like a good place to also call out the ability to retrieve existing type classes via `Puppet::Type.type :ec2_instance`.